src/main/java/org/traccar/helper/CoordinateUtil.java

```diff
@@ -0,0 +1,247 @@
+     *
+     * @param longitude Longitude
+     * @param latitude Latitude
+     * @return WGS84 coordinate
+     */
+    public static Coordinate gcj02ToWgs84(double longitude, double latitude) {
+        return new Coordinate(longitude, latitude).offset(offset(longitude, latitude, false));
+    }
+
+    /**
+     * Offset algorithm for WGS84 to GCJ-02 conversion (not exact)
+     *
+     * @param longitude Longitude
+     * @param latitude Latitude
+     * @param isPlus Forward offset or not: WGS84 to GCJ-02 use forward, otherwise use reverse.
+     * @return Offset coordinates
+     */
+    private static Coordinate offset(double longitude, double latitude, boolean isPlus) {
+        double dlng = transLng(longitude - 105.0, latitude - 35.0);
+        double dlat = transLat(longitude - 105.0, latitude - 35.0);
+
+        double magic = Math.sin(latitude / 180.0 * PI);
+        magic = 1 - CORRECTION_PARAM * magic * magic;
+        final double sqrtMagic = Math.sqrt(magic);
+
+        dlng = (dlng * 180.0) / (RADIUS / sqrtMagic * Math.cos(latitude / 180.0 * PI) * PI);
+        dlat = (dlat * 180.0) / ((RADIUS * (1 - CORRECTION_PARAM)) / (magic * sqrtMagic) * PI);
+
+        if (!isPlus) {
+            dlng = -dlng;
+            dlat = -dlat;
+        }
+
+        return new Coordinate(dlng, dlat);
+    }
+
+    /**
+     * Calculate longitude coordinates
+     *
+     * @param longitude Longitude
+     * @param latitude Latitude
+     * @return ret Calculated
+     */
+    private static double transLng(double longitude, double latitude) {
+        double ret = 300.0 + longitude + 2.0 * latitude
+                + 0.1 * longitude * longitude + 0.1 * longitude * latitude + 0.1 * Math.sqrt(Math.abs(longitude));
+        ret += (20.0 * Math.sin(6.0 * longitude * PI) + 20.0 * Math.sin(2.0 * longitude * PI)) * 2.0 / 3.0;
+        ret += (20.0 * Math.sin(longitude * PI) + 40.0 * Math.sin(longitude / 3.0 * PI)) * 2.0 / 3.0;
+        ret += (150.0 * Math.sin(longitude / 12.0 * PI) + 300.0 * Math.sin(longitude / 30.0 * PI)) * 2.0 / 3.0;
+        return ret;
+    }
+
+    /**
+     * Calculate latitude coordinates
+     *
+     * @param longitude Longitude
+     * @param latitude Latitude
+     * @return ret Calculated
+     */
+    private static double transLat(double longitude, double latitude) {
+        double ret = -100.0 + 2.0 * longitude + 3.0 * latitude + 0.2 * latitude * latitude + 0.1 * longitude * latitude
+                + 0.2 * Math.sqrt(Math.abs(longitude));
+        ret += (20.0 * Math.sin(6.0 * longitude * PI) + 20.0 * Math.sin(2.0 * longitude * PI)) * 2.0 / 3.0;
+        ret += (20.0 * Math.sin(latitude * PI) + 40.0 * Math.sin(latitude / 3.0 * PI)) * 2.0 / 3.0;
+        ret += (160.0 * Math.sin(latitude / 12.0 * PI) + 320 * Math.sin(latitude * PI / 30.0)) * 2.0 / 3.0;
+        return ret;
+    }
+
+
+    /**
+     * Coordinate latitude and longitude
+     */
+    public static class Coordinate implements Serializable {
```

Comment:
I think we can just use `Pair` class instead of this.

Reply:
At the beginning, I also thought of using "Pair". But then I referred to the code of "hutool" toolkit about coordinate system conversion, so I followed the code of this toolkit directly.

Reply:
I still prefer to use pair here.

Reply:
Just a reminder that this is still not addressed.